### ISSUE TITLE
Orphan types are exposed when using interfaces

### ISSUE THREAD

**Issue by tjoyal** [rmosolgo/graphql-ruby#4587]
**Describe the bug**

There appears to be an unexpected behavior where an orphan type implementing an interface ends up in the final schema.

As I understand it, it makes sense for the type to be made visible if the interface was a potential result of a visible field. However, in this repro case, this is not the case and the interface is not a response type of any of the current fields. The type gets exposed to the schema even when there is no path leading to it.

**Versions**

`graphql` version: 2.0.26
`sha`: `15cf025`

**Steps to reproduce**

I've created a small test case here: https://github.com/rmosolgo/graphql-ruby/compare/master...tjoyal:graphql-ruby:leak-hidden-types-with-interface

**Expected behavior**

Types that are unreachable shouldn't be exposed.

**Actual behavior**

Unreachable type is exposed.

**Additional context**

In typical setups, I could bypass the issue by duplicating the visibility flag on both the field and the type.

This is not ideal as it necessitates duplication of visibility definitions instead of setting it once at the base of the chain. I suspect some implementations out there could be exposing unused and sensitive types without their knowledge (e.g., schema references to unreleased features).

However, this becomes a larger nuisance when we try to re-use the same Ruby object for multiple GraphQL schemas, when we want to have distinct APIs while maintaining some consistency between them. In that scenario, it's not always possible to alter the reused Ruby object, and the original desire to set the visibility on the field itself is what we would prefer to do.

I haven't tested it, but I presume that union_memberships are also subject to this behavior.

https://github.com/rmosolgo/graphql-ruby/blob/15cf025a2fd34518529c95054a10d4be1985ff22/lib/graphql/schema/warden.rb#L295-L310

https://github.com/rmosolgo/graphql-ruby/blob/15cf025a2fd34518529c95054a10d4be1985ff22/lib/graphql/schema/warden.rb#L372-L377

**Comment by rmosolgo** [comment]
Hey, thanks for the detailed writeup. I definitely agree that this is unexpected (and undesirable) behavior. I don't think it was added on purpose, but I remember that there were some known edge cases like this (eg #2645) that were just hard to work out. 

I tried working up a fix in #4589, but it looks like it broke some other functionality. I'll see if I can find something that works for _all_ the tests!
